metapool/sample_sheet.py

```diff
@@ -826,3 +824,149 @@ def sample_sheet_to_dataframe(sheet):
     out.drop(columns='Sample_Project', inplace=True)
     out.sort_values(by='sample_well', inplace=True)
     return out.set_index('sample_id')
+
+
+def contains_replicates(sheet):
+    """Returns True if samples in sheet include one or more replicates.
+
+    Parameters
+    ----------
+    sheet: sample_sheet.KLSampleSheet
+        Object from where to extract the data.
+
+    Returns
+    -------
+    bool
+        True if samples in sheet include one or more replicates.
+    """
+    if 'contains_replicates' in sheet.Bioinformatics.columns:
+        for index, row in sheet.Bioinformatics.iterrows():
+            if row['contains_replicates'].lower() == 'true':
+                return True
+
+    # legacy sample-sheet does not handle replicates or no replicates were
+    # found.
+    return False
+
+
+def _demux_sample_sheet(sheet, project_name):
+    df = sample_sheet_to_dataframe(sheet)
+
+    if project_name not in set(df['sample_project']):
+        raise ValueError(f"Project '{project_name}' is not defined in "
+                         "sample-sheet")
+
+    # filter out samples not associated w/this project
+    df = df.loc[df['sample_project'] == project_name]
+
+    # use PlateReplication object to convert each sample's 384 well location
+    # into a 96-well location + quadrant. Since replication is performed at
+    # the plate-level, this will identify which replicants belong in which
+    # new sample-sheet.
+    plate = PlateReplication(None)
+
+    df['quad'] = df.apply(lambda row: plate.get_96_well_location_and_quadrant(
+        row.destination_well_384)[0], axis=1)
+
+    gb = df.groupby('quad')
+
+    results = [gb.get_group(grp).drop(['quad'], axis=1) for grp in gb.groups]
+
+    # clean-up each dataframe before returning to the user.
+    for result in results:
+        result['Sample_ID'] = result.index
+        result.reset_index(inplace=True)
+
+    return results
```

Comment:
Similar to the other 2 comments about using group_by correctly.

Reply:
I'll have to put more thought into this one. I'm not sure yet how to do all of that in fewer lines.

Reply:
I believe my new solution looks better - I removed the use of groupby entirely and simply create new data frames according to unique values found in the quad column.